libbeat/cfgfile/list.go

```diff
@@ -0,0 +1,133 @@
+package cfgfile
+
+import (
+	"sync"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// RunnerList implements a reloadable.List of Runners
+type RunnerList struct {
+	runners  map[uint64]Runner
+	mutex    sync.Mutex
+	factory  RunnerFactory
+	pipeline beat.Pipeline
+}
+
+// NewRunnerList builds and returns a RunnerList
+func NewRunnerList(factory RunnerFactory, pipeline beat.Pipeline) *RunnerList {
+	return &RunnerList{
+		runners:  map[uint64]Runner{},
+		factory:  factory,
+		pipeline: pipeline,
+	}
+}
+
+// Reload the list of runners to match the given state
+// TODO return list of errors (even we don't stop on them)
+func (r *RunnerList) Reload(configs []*common.Config) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	startList := map[uint64]*common.Config{}
+	stopList := r.copyRunnerList()
+
+	debugf("Starting reload procedure, current runners: %d", len(stopList))
+
+	// diff current & desired state, create action lists
+	for _, config := range configs {
+		hash, err := HashConfig(config)
+		if err != nil {
+			logp.Err("Unable to hash given config: %v", err)
+			continue
+		}
+
+		if _, ok := stopList[hash]; ok {
+			delete(stopList, hash)
+		} else {
+			startList[hash] = config
+		}
+	}
+
+	// Stop removed runners
+	for hash, runner := range stopList {
+		debugf("Stopping runner: %s", runner)
+		delete(r.runners, hash)
+		go runner.Stop()
+	}
+
+	// Start new runners
+	for hash, config := range startList {
+		runner, err := r.factory.Create(r.pipeline, config, nil)
+		if err != nil {
+			debugf("Error creating runner from config: %s", err)
+			continue
+		}
+
+		debugf("Starting runner: %s", runner)
+		r.runners[hash] = runner
+		runner.Start()
+	}
+
+	return nil
+}
+
+// Stop all runners
+func (r *RunnerList) Stop() {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	for h, runner := range r.runners {
+		debugf("Stopping runner: %s", runner)
+		delete(r.runners, h)
+		runner.Stop()
```

Comment:
Before runners were stopped in parallel. This is important to speed up the shutdown process especially with a large amount of modules or prospectors.

Reply:
makes sense, will update the code to do that